tests/unit/s2n_server_hello_retry_test.c

```diff
@@ -37,8 +40,45 @@ const uint8_t COMPRESSION_METHOD_SIZE = 1;
 struct client_hello_context {
     int invocations;
     s2n_client_hello_cb_mode mode;
+    bool mark_done;
+    bool enable_poll;
 };
 
+int s2n_negotiate_poll_hello_retry(struct s2n_connection *server_conn,
+                struct s2n_connection *client_conn,
+                struct client_hello_context *client_hello_ctx)
+{
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(client_conn, &blocked), S2N_ERR_IO_BLOCKED);
+
+    int expected_invocation = 0;
+
+    /* if polling is enabled then confirm that the callback is incremented each time */
+    if (client_hello_ctx->enable_poll) {
+        do {
+            /* invocation should increase each time s2n_negotiate is called */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(server_conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+            expected_invocation++;
+            EXPECT_EQUAL(client_hello_ctx->invocations, expected_invocation);
+        } while (expected_invocation < 2);
+    }
```

Comment:
Since you can easily do this now, I'd run this more than twice. Checking something that should work indefinitely only twice makes me nervous: that could try initial logic once, reentry logic once, and never reentry logic again, introducing the possibility that the reentry logic mutates state in an unsustainable way. I'd repeat at least 3 times, probably 5 or 10 since it costs basically nothing to add more iterations.

Although s2n_self_talk_client_hello_cb_test.c might be a better place to worry about more iterations.